Add compatability to tinker swords

Raycoms opened the pull request:
And make guards smarter,
Also let guards focus last attackers
Don't let players mark patrol points outside of the patrol  area

Closes #279 
Closes #429 

# Changes proposed in this pull request:
-
-
-

Review please

Raycoms committed:
Add compatability to tinker swords

And make guards smarter,
Also let guards focus last attackers
Don't let players mark patrol points outside of the patrol  area

MinecoloniesSonar commented on src/main/java/com/minecolonies/compatibility/Compatibility.java:
```diff
@@ -1,12 +1,14 @@
 import net.minecraft.item.ItemStack;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
+import slimeknights.tconstruct.library.utils.ToolHelper;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Remove this unused import 'slimeknights.tconstruct.library.utils.ToolHelper'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

marvin-bitterlich replied:
this will crash!!!!

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:
```diff
@@ -37,6 +37,11 @@
 public abstract class AbstractEntityAIBasic<J extends AbstractJob> extends AbstractAISkeleton<J>
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 25 to the maximum authorized 20 or less. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') class "AbstractEntityAIBasic" has 61 methods, which is greater than the 35 authorized. Split it into smaller classes. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)

MinecoloniesSonar commented:
SonarQube analysis reported 82 issues
* ![MAJOR][MAJOR] 26 major
* ![MINOR][MINOR] 51 minor
* ![INFO][INFO] 5 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Compatibility.java#L32](https://github.com/Minecolonies/minecolonies/blob/cd92c2cc824fd228c229879efaed4ab38b926321/src/main/java/com/minecolonies/compatibility/Compatibility.java#L32): Rename this method to start with "is" or "has". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2047)
1. ![MAJOR][MAJOR] [Compatibility.java#L32](https://github.com/Minecolonies/minecolonies/blob/cd92c2cc824fd228c229879efaed4ab38b926321/src/main/java/com/minecolonies/compatibility/Compatibility.java#L32): Remove this unused method parameter "tool". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1172)
1. ![MAJOR][MAJOR] [BuildingGuardTower.java#L549](https://github.com/Minecolonies/minecolonies/blob/cd92c2cc824fd228c229879efaed4ab38b926321/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingGuardTower.java#L549): Reduce this class from 58 to the maximum allowed 25 or externalize it in a public class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [EntityCitizen.java](https://github.com/Minecolonies/minecolonies/blob/cd92c2cc824fd228c229879efaed4ab38b926321/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java): This file has 1.170 lines, which is greater than 750 authorized. Split it into smaller files. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L62](https://github.com/Minecolonies/minecolonies/blob/cd92c2cc824fd228c229879efaed4ab38b926321/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L62): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 60 to the maximum authorized 20 or less. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L62](https://github.com/Minecolonies/minecolonies/blob/cd92c2cc824fd228c229879efaed4ab38b926321/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L62): class "EntityCitizen" has 92 methods, which is greater than the 35 authorized. Split it into smaller classes. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L287](https://github.com/Minecolonies/minecolonies/blob/cd92c2cc824fd228c229879efaed4ab38b926321/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L287): Assign this magic number 3 to a well-named constant, and use the constant instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L288](https://github.com/Minecolonies/minecolonies/blob/cd92c2cc824fd228c229879efaed4ab38b926321/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L288): Assign this magic number 4 to a well-named constant, and use the constant instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L289](https://github.com/Minecolonies/minecolonies/blob/cd92c2cc824fd228c229879efaed4ab38b926321/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L289): Assign this magic number 4 to a well-named constant, and use the constant instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L290](https://github.com/Minecolonies/minecolonies/blob/cd92c2cc824fd228c229879efaed4ab38b926321/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L290): Assign this magic number 5 to a well-named constant, and use the constant instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

marvin-bitterlich commented on src/main/java/com/minecolonies/compatibility/tinkers/TinkersWeaponHelper.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Class to check if certain tinkers items serve as weapons for the guards.
+ */
+public final class TinkersWeaponHelper
```
this is the working class. We need a fallback class and the @Optional annotations. This will not work!

marchermans replied:
Exactly.
There are even better implementations and i am working on those, but @Optional is the minimal requirement

marvin-bitterlich reviewed (changes requested):
This will not work with tinkers absent! Check out how we did the optional please

Raycoms committed:
add compatability

MinecoloniesSonar commented on src/main/java/com/minecolonies/compatibility/tinkers/TinkersWeaponProxy.java:
```diff
@@ -0,0 +1,32 @@
+     * @param stack the stack to check for.
+     * @return true if so.
+     */
+    public boolean isTinkersWeapon(@NotNull final ItemStack stack)
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove this unused method parameter "stack". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1172)

MinecoloniesSonar commented on src/main/java/com/minecolonies/compatibility/tinkers/TinkersWeaponProxy.java:
```diff
@@ -0,0 +1,32 @@
+
+import net.minecraft.item.ItemStack;
+import org.jetbrains.annotations.NotNull;
+import slimeknights.tconstruct.library.tools.SwordCore;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Remove this unused import 'slimeknights.tconstruct.library.tools.SwordCore'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

MinecoloniesSonar commented on src/main/java/com/minecolonies/compatibility/tinkers/TinkersWeaponProxy.java:
```diff
@@ -0,0 +1,32 @@
+import net.minecraft.item.ItemStack;
+import org.jetbrains.annotations.NotNull;
+import slimeknights.tconstruct.library.tools.SwordCore;
+import slimeknights.tconstruct.library.utils.ToolHelper;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Remove this unused import 'slimeknights.tconstruct.library.utils.ToolHelper'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

MinecoloniesSonar commented on src/main/java/com/minecolonies/compatibility/tinkers/TinkersWeaponProxy.java:
```diff
@@ -0,0 +1,32 @@
+     * @param stack the stack.
+     * @return the attack damage.
+     */
+    public double getAttackDamage(@NotNull final ItemStack stack)
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove this unused method parameter "stack". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1172)

MinecoloniesSonar commented on src/main/java/com/minecolonies/compatibility/tinkers/TinkersWeaponProxy.java:
```diff
@@ -0,0 +1,32 @@
+     */
+    public double getAttackDamage(@NotNull final ItemStack stack)
+    {
+        return 0;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Remove this method and declare a constant for this value. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3400)

MinecoloniesSonar commented:
SonarQube analysis reported 88 issues
* ![MAJOR][MAJOR] 29 major
* ![MINOR][MINOR] 54 minor
* ![INFO][INFO] 5 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Compatibility.java#L32](https://github.com/Minecolonies/minecolonies/blob/432767c3e6cef1e3de69ba3c68b10b8ba46fc600/src/main/java/com/minecolonies/compatibility/Compatibility.java#L32): Rename this method to start with "is" or "has". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2047)
1. ![MAJOR][MAJOR] [Compatibility.java#L32](https://github.com/Minecolonies/minecolonies/blob/432767c3e6cef1e3de69ba3c68b10b8ba46fc600/src/main/java/com/minecolonies/compatibility/Compatibility.java#L32): Remove this unused method parameter "tool". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1172)
1. ![MAJOR][MAJOR] [ToolBrokenCheck.java#L10](https://github.com/Minecolonies/minecolonies/blob/432767c3e6cef1e3de69ba3c68b10b8ba46fc600/src/main/java/com/minecolonies/compatibility/tinkers/ToolBrokenCheck.java#L10): Add a private constructor to hide the implicit public one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1118)
1. ![MAJOR][MAJOR] [BuildingGuardTower.java#L549](https://github.com/Minecolonies/minecolonies/blob/432767c3e6cef1e3de69ba3c68b10b8ba46fc600/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingGuardTower.java#L549): Reduce this class from 58 to the maximum allowed 25 or externalize it in a public class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [EntityCitizen.java](https://github.com/Minecolonies/minecolonies/blob/432767c3e6cef1e3de69ba3c68b10b8ba46fc600/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java): This file has 1.170 lines, which is greater than 750 authorized. Split it into smaller files. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L62](https://github.com/Minecolonies/minecolonies/blob/432767c3e6cef1e3de69ba3c68b10b8ba46fc600/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L62): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 60 to the maximum authorized 20 or less. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L62](https://github.com/Minecolonies/minecolonies/blob/432767c3e6cef1e3de69ba3c68b10b8ba46fc600/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L62): class "EntityCitizen" has 92 methods, which is greater than the 35 authorized. Split it into smaller classes. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L287](https://github.com/Minecolonies/minecolonies/blob/432767c3e6cef1e3de69ba3c68b10b8ba46fc600/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L287): Assign this magic number 3 to a well-named constant, and use the constant instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L288](https://github.com/Minecolonies/minecolonies/blob/432767c3e6cef1e3de69ba3c68b10b8ba46fc600/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L288): Assign this magic number 4 to a well-named constant, and use the constant instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L289](https://github.com/Minecolonies/minecolonies/blob/432767c3e6cef1e3de69ba3c68b10b8ba46fc600/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L289): Assign this magic number 4 to a well-named constant, and use the constant instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

Raycoms committed:
imports

MinecoloniesSonar commented:
SonarQube analysis reported 85 issues
* ![MAJOR][MAJOR] 29 major
* ![MINOR][MINOR] 51 minor
* ![INFO][INFO] 5 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Compatibility.java#L31](https://github.com/Minecolonies/minecolonies/blob/54be4053d5ad74497629e025401aa97b30b10cf3/src/main/java/com/minecolonies/compatibility/Compatibility.java#L31): Rename this method to start with "is" or "has". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2047)
1. ![MAJOR][MAJOR] [Compatibility.java#L31](https://github.com/Minecolonies/minecolonies/blob/54be4053d5ad74497629e025401aa97b30b10cf3/src/main/java/com/minecolonies/compatibility/Compatibility.java#L31): Remove this unused method parameter "tool". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1172)
1. ![MAJOR][MAJOR] [ToolBrokenCheck.java#L10](https://github.com/Minecolonies/minecolonies/blob/54be4053d5ad74497629e025401aa97b30b10cf3/src/main/java/com/minecolonies/compatibility/tinkers/ToolBrokenCheck.java#L10): Add a private constructor to hide the implicit public one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1118)
1. ![MAJOR][MAJOR] [BuildingGuardTower.java#L549](https://github.com/Minecolonies/minecolonies/blob/54be4053d5ad74497629e025401aa97b30b10cf3/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingGuardTower.java#L549): Reduce this class from 58 to the maximum allowed 25 or externalize it in a public class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [EntityCitizen.java](https://github.com/Minecolonies/minecolonies/blob/54be4053d5ad74497629e025401aa97b30b10cf3/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java): This file has 1.170 lines, which is greater than 750 authorized. Split it into smaller files. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L62](https://github.com/Minecolonies/minecolonies/blob/54be4053d5ad74497629e025401aa97b30b10cf3/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L62): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 60 to the maximum authorized 20 or less. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L62](https://github.com/Minecolonies/minecolonies/blob/54be4053d5ad74497629e025401aa97b30b10cf3/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L62): class "EntityCitizen" has 92 methods, which is greater than the 35 authorized. Split it into smaller classes. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L287](https://github.com/Minecolonies/minecolonies/blob/54be4053d5ad74497629e025401aa97b30b10cf3/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L287): Assign this magic number 3 to a well-named constant, and use the constant instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L288](https://github.com/Minecolonies/minecolonies/blob/54be4053d5ad74497629e025401aa97b30b10cf3/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L288): Assign this magic number 4 to a well-named constant, and use the constant instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L289](https://github.com/Minecolonies/minecolonies/blob/54be4053d5ad74497629e025401aa97b30b10cf3/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L289): Assign this magic number 4 to a well-named constant, and use the constant instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

Raycoms committed:
Let all workers gain xp, and improve imports

Guards only gain xp by monster xp drops now.
Rangers will get that when they gather

MinecoloniesSonar commented:
SonarQube analysis reported 96 issues
* ![MAJOR][MAJOR] 38 major
* ![MINOR][MINOR] 52 minor
* ![INFO][INFO] 6 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Compatibility.java#L31](https://github.com/Minecolonies/minecolonies/blob/15d0597e21d8119ad4492a59b880003027eed110/src/main/java/com/minecolonies/compatibility/Compatibility.java#L31): Rename this method to start with "is" or "has". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2047)
1. ![MAJOR][MAJOR] [Compatibility.java#L31](https://github.com/Minecolonies/minecolonies/blob/15d0597e21d8119ad4492a59b880003027eed110/src/main/java/com/minecolonies/compatibility/Compatibility.java#L31): Remove this unused method parameter "tool". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1172)
1. ![MAJOR][MAJOR] [ToolBrokenCheck.java#L10](https://github.com/Minecolonies/minecolonies/blob/15d0597e21d8119ad4492a59b880003027eed110/src/main/java/com/minecolonies/compatibility/tinkers/ToolBrokenCheck.java#L10): Add a private constructor to hide the implicit public one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1118)
1. ![MAJOR][MAJOR] [BuildingGuardTower.java#L549](https://github.com/Minecolonies/minecolonies/blob/15d0597e21d8119ad4492a59b880003027eed110/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingGuardTower.java#L549): Reduce this class from 58 to the maximum allowed 25 or externalize it in a public class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [AbstractCitizensCommands.java#L44](https://github.com/Minecolonies/minecolonies/blob/15d0597e21d8119ad4492a59b880003027eed110/src/main/java/com/minecolonies/coremod/commands/AbstractCitizensCommands.java#L44): Refactor this method to reduce its Cognitive Complexity from 21 to the 20 allowed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [RandomTeleportCommand.java#L130](https://github.com/Minecolonies/minecolonies/blob/15d0597e21d8119ad4492a59b880003027eed110/src/main/java/com/minecolonies/coremod/commands/RandomTeleportCommand.java#L130): Reduce the total number of break and continue statements in this loop to use at most one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS135)
1. ![MAJOR][MAJOR] [EntityCitizen.java](https://github.com/Minecolonies/minecolonies/blob/15d0597e21d8119ad4492a59b880003027eed110/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java): This file has 1.171 lines, which is greater than 750 authorized. Split it into smaller files. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L62](https://github.com/Minecolonies/minecolonies/blob/15d0597e21d8119ad4492a59b880003027eed110/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L62): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 60 to the maximum authorized 20 or less. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L62](https://github.com/Minecolonies/minecolonies/blob/15d0597e21d8119ad4492a59b880003027eed110/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L62): class "EntityCitizen" has 92 methods, which is greater than the 35 authorized. Split it into smaller classes. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L287](https://github.com/Minecolonies/minecolonies/blob/15d0597e21d8119ad4492a59b880003027eed110/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L287): Assign this magic number 3 to a well-named constant, and use the constant instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

Raycoms committed:
Fix citizen data crash

MinecoloniesSonar commented:
SonarQube analysis reported 96 issues
* ![MAJOR][MAJOR] 38 major
* ![MINOR][MINOR] 52 minor
* ![INFO][INFO] 6 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Compatibility.java#L31](https://github.com/Minecolonies/minecolonies/blob/743af321963b30105f63a89f3e2fc03ad233acbd/src/main/java/com/minecolonies/compatibility/Compatibility.java#L31): Rename this method to start with "is" or "has". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2047)
1. ![MAJOR][MAJOR] [Compatibility.java#L31](https://github.com/Minecolonies/minecolonies/blob/743af321963b30105f63a89f3e2fc03ad233acbd/src/main/java/com/minecolonies/compatibility/Compatibility.java#L31): Remove this unused method parameter "tool". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1172)
1. ![MAJOR][MAJOR] [ToolBrokenCheck.java#L10](https://github.com/Minecolonies/minecolonies/blob/743af321963b30105f63a89f3e2fc03ad233acbd/src/main/java/com/minecolonies/compatibility/tinkers/ToolBrokenCheck.java#L10): Add a private constructor to hide the implicit public one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1118)
1. ![MAJOR][MAJOR] [BuildingGuardTower.java#L549](https://github.com/Minecolonies/minecolonies/blob/743af321963b30105f63a89f3e2fc03ad233acbd/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingGuardTower.java#L549): Reduce this class from 58 to the maximum allowed 25 or externalize it in a public class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [AbstractCitizensCommands.java#L44](https://github.com/Minecolonies/minecolonies/blob/743af321963b30105f63a89f3e2fc03ad233acbd/src/main/java/com/minecolonies/coremod/commands/AbstractCitizensCommands.java#L44): Refactor this method to reduce its Cognitive Complexity from 21 to the 20 allowed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [RandomTeleportCommand.java#L130](https://github.com/Minecolonies/minecolonies/blob/743af321963b30105f63a89f3e2fc03ad233acbd/src/main/java/com/minecolonies/coremod/commands/RandomTeleportCommand.java#L130): Reduce the total number of break and continue statements in this loop to use at most one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS135)
1. ![MAJOR][MAJOR] [EntityCitizen.java](https://github.com/Minecolonies/minecolonies/blob/743af321963b30105f63a89f3e2fc03ad233acbd/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java): This file has 1.178 lines, which is greater than 750 authorized. Split it into smaller files. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L62](https://github.com/Minecolonies/minecolonies/blob/743af321963b30105f63a89f3e2fc03ad233acbd/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L62): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 60 to the maximum authorized 20 or less. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L62](https://github.com/Minecolonies/minecolonies/blob/743af321963b30105f63a89f3e2fc03ad233acbd/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L62): class "EntityCitizen" has 92 methods, which is greater than the 35 authorized. Split it into smaller classes. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L287](https://github.com/Minecolonies/minecolonies/blob/743af321963b30105f63a89f3e2fc03ad233acbd/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L287): Assign this magic number 3 to a well-named constant, and use the constant instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

Raycoms commented:
Fixed everything @Kostronor

marchermans commented:
Yep much better.

marvin-bitterlich commented:
:shipit: 

[![Approved with PullApprove](https://img.shields.io/badge/maintainers-approved-brightgreen.svg)](https://pullapprove.com/Minecolonies/minecolonies/pull-request/844/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)
